File: Source/OldCode/Archive/2005 TVA Code Library/Version/CS/TVA.Core_CS/Net/Ftp/Directory.cs
using System.Diagnostics;
using System.Linq;
using System.Data;
using System.Collections;
using Microsoft.VisualBasic;
using System.Collections.Generic;
using System;
using System.IO;
using System.Text;
using System.Net.Sockets;
using System.Text.RegularExpressions;

// James Ritchie Carroll - 2003


namespace TVA
{
    namespace Net
    {
        namespace Ftp
        {


            public interface IFile
            {

                Directory Parent
                {
                    get;
                }
                string Name
                {
                    get;
                }
                string FullPath
                {
                    get;
                }
                bool IsFile
                {
                    get;
                }
                bool IsDirectory
                {
                    get;
                }
                long Size
                {
                    get;
                    set;
                }
                string Permission
                {
                    get;
                    set;
                }
                DateTime TimeStamp
                {
                    get;
                    set;
                }

            }

            internal class TimeStampParser
            {


                public enum RawDataStyle
                {
                    UnixDate,
                    UnixDateTime,
                    DosDateTime,
                    Undetermined
                }

                public string RawValue;
                public RawDataStyle Style;

                public TimeStampParser()
                {

                    Style = RawDataStyle.Undetermined;

                }

                public TimeStampParser(string RawValue, RawDataStyle Style)
                {

                    this.RawValue = RawValue;
                    this.Style = Style;

                }

                public DateTime Value
                {
                    get
                    {
                        if (RawValue.Length > 0)
                        {
                            try
                            {
                                switch (Style)
                                {
                                    case RawDataStyle.UnixDate:
                                        return DateTime.Parse(RawValue);
                                    case RawDataStyle.UnixDateTime:
                                        string[] sa = RawValue.Split(' ');
                                        return System.Convert.ToDateTime(sa[0] + " " + sa[1] + " " + DateAndTime.Year(DateTime.Now) + " " + sa[2]);
                                    case RawDataStyle.DosDateTime:
                                        return DateTime.Parse(RawValue);
                                    default:
                                        return DateTime.Parse(RawValue);
                                }
                            }
                            catch
                            {
                                return DateTime.MinValue;
                            }
                        }
                        else
                        {
                            return DateTime.MinValue;
                        }
                    }
                }

            }

            internal class ItemInfo
            {


                public string Name;
                public string FullPath;
                public string Permission;
                public bool IsDirectory;
                public long Size;
                public TimeStampParser TimeStamp = new TimeStampParser();

            }

            public class Directory : IFile, IComparable
            {



                private SessionConnected m_session;
                private Directory m_parent;
                private string m_name;
                private string m_fullPath;
                private Dictionary<string, Directory> m_subDirectories;
                private Dictionary<string, System.IO.File> m_files;
                private bool m_caseInsensitive;
                private long m_size;
                private string m_permission;
                private DateTime m_timestamp;

                private static Regex m_UnixListLineStyle1 = new Regex("(?<dir>[\\-d])(?<permission>([\\-r][\\-w][\\-xs]){3})\\s+\\d+\\s+\\w+\\s+\\w+\\s+(?<size>\\d+)\\s+(?<timestamp>\\w+\\s+\\d+\\s+\\d{4})\\s+(?<name>.+)");
                private static Regex m_UnixListLineStyle2 = new Regex("(?<dir>[\\-d])(?<permission>([\\-r][\\-w][\\-xs]){3})\\s+\\d+\\s+\\d+\\s+(?<size>\\d+)\\s+(?<timestamp>\\w+\\s+\\d+\\s+\\d{4})\\s+(?<name>.+)");
                private static Regex m_UnixListLineStyle3 = new Regex("(?<dir>[\\-d])(?<permission>([\\-r][\\-w][\\-xs]){3})\\s+\\d+\\s+\\d+\\s+(?<size>\\d+)\\s+(?<timestamp>\\w+\\s+\\d+\\s+\\d+:\\d+)\\s+(?<name>.+)");
                private static Regex m_DosListLineStyle1 = new Regex("(?<timestamp>\\d{2}\\-\\d{2}\\-\\d{2}\\s+\\d{2}:\\d{2}[Aa|Pp][mM])\\s+(?<dir>\\<\\w+\\>){0,1}(?<size>\\d+){0,1}\\s+(?<name>.+)"); // IIS FTP Service
                private static Regex m_DosListLineStyle2 = new Regex("(?<dir>[\\-d])(?<permission>([\\-r][\\-w][\\-xs]){3})\\s+\\d+\\s+\\w+\\s+\\w+\\s+(?<size>\\d+)\\s+(?<timestamp>\\w+\\s+\\d+\\s+\\d+:\\d+)\\s+(?<name>.+)"); // IIS FTP Service in Unix Mode

                public delegate void DirectoryListLineScanEventHandler(string Line);
                private DirectoryListLineScanEventHandler DirectoryListLineScanEvent;

                public event DirectoryListLineScanEventHandler DirectoryListLineScan
                {
                    add
                    {
                        DirectoryListLineScanEvent = (DirectoryListLineScanEventHandler)System.Delegate.Combine(DirectoryListLineScanEvent, value);
                    }
                    remove
                    {
                        DirectoryListLineScanEvent = (DirectoryListLineScanEventHandler)System.Delegate.Remove(DirectoryListLineScanEvent, value);
                    }
                }

                public delegate void DirectoryScanExceptionEventHandler(Exception ex);
                private DirectoryScanExceptionEventHandler DirectoryScanExceptionEvent;

                public event DirectoryScanExceptionEventHandler DirectoryScanException
                {
                    add
                    {
                        DirectoryScanExceptionEvent = (DirectoryScanExceptionEventHandler)System.Delegate.Combine(DirectoryScanExceptionEvent, value);
                    }
                    remove
                    {
                        DirectoryScanExceptionEvent = (DirectoryScanExceptionEventHandler)System.Delegate.Remove(DirectoryScanExceptionEvent, value);
                    }
                }


                internal Directory(SessionConnected s, bool CaseInsensitive, string fullPath)
                {

                    m_session = s;
                    m_parent = null;
                    m_caseInsensitive = CaseInsensitive;

                    if (fullPath.Substring(fullPath.Length - 1, 1) == "/")
                    {
                        fullPath = fullPath.Substring(0, fullPath.Length - 1);
                    }

                    if (fullPath.Length == 0)
                    {
                        m_name = m_fullPath == "/";
                    }
                    else
                    {
                        string[] directories = fullPath.Split('/');
                        m_name = directories[directories.Length - 1];
                        m_fullPath = fullPath + "/";
                    }

                }

                internal Directory(SessionConnected s, Directory parent, bool CaseInsensitive, ItemInfo info)
                {

                    m_session = s;
                    m_parent = parent;
                    m_caseInsensitive = CaseInsensitive;

                    if (info.Name.Length > 0)
                    {
                        m_name = info.Name;
                        if (parent == null)
                        {
                            m_fullPath = m_name + "/";
                        }
                        else
                        {
                            m_fullPath = parent.FullPath + m_name + "/";
                        }
                    }
                    else
                    {
                        m_name = m_fullPath == "/";
                    }

                    m_size = info.Size;
                    m_permission = info.Permission;
                    m_timestamp = info.TimeStamp.Value;

                }

                public bool CaseInsensitive
                {
                    get
                    {
                        return m_caseInsensitive;
                    }
                    set
                    {
                        m_caseInsensitive = value;
                        Refresh();
                    }
                }

                public string Name
                {
                    get
                    {
                        return m_name;
                    }
                }

                public string FullPath
                {
                    get
                    {
                        return m_fullPath;
                    }
                }

                public bool IsFile
                {
                    get
                    {
                        return false;
                    }
                }

                public bool IsDirectory
                {
                    get
                    {
                        return true;
                    }
                }

                public long Size
                {
                    get
                    {
                        return m_size;
                    }
                    set
                    {
                        m_size = value;
                    }
                }

                public string Permission
                {
                    get
                    {
                        return m_permission;
                    }
                    set
                    {
                        m_permission = value;
                    }
                }

                public DateTime TimeStamp
                {
                    get
                    {
                        return m_timestamp;
                    }
                    set
                    {
                        m_timestamp = value;
                    }
                }

                public Directory Parent
                {
                    get
                    {
                        if (Strings.StrComp(m_fullPath, m_session.RootDirectory.m_fullPath, (m_caseInsensitive ? CompareMethod.Text : CompareMethod.Binary)) == 0)
                        {
                            return null;
                        }

                        // If we don't have a reference to parent directory, we try to derive it...
                        if (m_parent == null)
                        {
                            CheckSessionCurrentDirectory();

                            StringBuilder parentPath = new StringBuilder();
                            var fullPath = m_session.ControlChannel.PWD();

                            if (fullPath.Substring(fullPath.Length - 1, 1) != "/")
                            {
                                fullPath += "/";
                            }

                            string[] paths = fullPath.Split("/");
                            int i;

                            for (i = 0; i <= paths.Length - 2; i++)
                            {
                                if (paths[i].Length == 0)
                                {
                                    parentPath.Append("/");
                                }
                                else
                                {
                                    parentPath.Append(paths[i]);
                                    parentPath.Append("/");
                                }
                            }

                            Directory parentDir = new Directory(m_session, m_caseInsensitive, parentPath.ToString());

                            if (Strings.StrComp(parentDir.m_fullPath, m_session.RootDirectory.m_fullPath, (m_caseInsensitive ? CompareMethod.Text : CompareMethod.Binary)) == 0)
                            {
                                m_parent = m_session.RootDirectory;
                            }
                            else
                            {
                                m_parent = parentDir;
                            }
                        }

                        return m_parent;
                    }
                }

                public Dictionary<string, Directory> SubDirectories
                {
                    get
                    {
                        InitHashtable();
                        return m_subDirectories.Values;
                    }
                }

                public Dictionary<string, System.IO.File> Files
                {
                    get
                    {
                        InitHashtable();
                        return m_files.Values;
                    }
                }

                public File FindFile(string fileName)
                {

                    InitHashtable();

                    return m_files(fileName);

                }

                public Directory FindSubDirectory(string dirName)
                {

                    InitHashtable();

                    return m_subDirectories(dirName);

                }

                public void PutFile(string localFile)
                {

                    PutFile(localFile, null);

                }

                public void PutFile(string localFile, string remoteFile)
                {

                    CheckSessionCurrentDirectory();

                    FileInfo fi = new FileInfo(localFile);

                    if (remoteFile == null)
                    {
                        remoteFile = fi.Name;
                    }

                    FileTransferer transfer = new FileTransferer(this, localFile, remoteFile, fi.Length, TransferDirection.Upload);

                    transfer.StartTransfer();

                }

                public void GetFile(string remoteFile)
                {

                    GetFile(remoteFile, remoteFile);

                }

                public void GetFile(string localFile, string remoteFile)
                {

                    InitHashtable();

                    System.IO.File file = m_files(remoteFile);

                    if (file == null)
                    {
                        throw (new FileNotFoundException(remoteFile));
                    }

                    FileTransferer transfer = new FileTransferer(this, localFile, remoteFile, File.Size, TransferDirection.Download);

                    transfer.StartTransfer();

                }

                public void BeginPutFile(string localFile)
                {

                    BeginPutFile(localFile, null);

                }

                public void BeginPutFile(string localFile, string remoteFile)
                {

                    CheckSessionCurrentDirectory();

                    FileInfo fi = new FileInfo(localFile);

                    if (remoteFile == null)
                    {
                        remoteFile = fi.Name;
                    }

                    FileTransferer transfer = new FileTransferer(this, localFile, remoteFile, fi.Length, TransferDirection.Upload);

                    transfer.StartAsyncTransfer();

                }

                public void BeginGetFile(string remoteFile)
                {

                    BeginGetFile(remoteFile, remoteFile);

                }

                public void BeginGetFile(string localFile, string remoteFile)
                {

                    InitHashtable();

                    File file = m_files(remoteFile);

                    if (file == null)
                    {
                        throw (new FileNotFoundException(remoteFile));
                    }

                    FileTransferer transfer = new FileTransferer(this, localFile, remoteFile, File.Size, TransferDirection.Download);

                    transfer.StartAsyncTransfer();

                }

                public void RemoveFile(string fileName)
                {

                    CheckSessionCurrentDirectory();

                    m_session.ControlChannel.DELE(fileName);

                    m_files.Remove(fileName);

                }

                public void RemoveSubDir(string dirName)
                {

                    CheckSessionCurrentDirectory();

                    m_session.ControlChannel.RMD(dirName);

                    m_subDirectories.Remove(dirName);

                }

                public File CreateFile(string newFileName)
                {

                    DataStream stream = CreateFileStream(newFileName);

                    stream.Close();

                    return m_files(newFileName);

                }

                public OutputDataStream CreateFileStream(string newFileName)
                {

                    InitHashtable();

                    DataStream stream = m_session.ControlChannel.GetPassiveDataStream(TransferDirection.Upload);

                    try
                    {
                        m_session.ControlChannel.STOR(newFileName);

                        File newFile = new File(this, newFileName);

                        m_files(newFileName) = newFile;

                        return ((OutputDataStream)stream);
                    }
                    catch
                    {
                        stream.Close();
                        throw;
                    }

                }

                public void Refresh()
                {

                    ClearItems();
                    InitHashtable();

                }

                internal void ClearItems()
                {

                    m_subDirectories = null;
                    m_files = null;

                }

                internal SessionConnected Session
                {
                    get
                    {
                        return m_session;
                    }
                }

                internal void CheckSessionCurrentDirectory()
                {

                    if (m_session.CurrentDirectory.m_fullPath != m_fullPath)
                    {
                        throw (new InvalidOperationException(m_fullPath + " is not current directory."));
                    }

                }

                private void LoadDirectoryItems()
                {

                    if (m_session.CurrentDirectory != this)
                    {
                        throw (new InvalidOperationException(m_name + " is not current active directory"));
                    }

                    Queue lineQueue = m_session.ControlChannel.List(false);
                    ItemInfo info;


                    foreach (string line in lineQueue)
                    {
                        // We allow users to inspect FTP lineQueue if desired...
                        if (DirectoryListLineScanEvent != null)
                            DirectoryListLineScanEvent(line);

                        try
                        {
                            info = new ItemInfo();
                            if (ParseListLine(line, info))
                            {
                                if (info.IsDirectory)
                                {
                                    m_subDirectories.Add(info.Name, new Directory(m_session, this, m_caseInsensitive, info));
                                }
                                else
                                {
                                    m_files.Add(info.Name, new File(this, info));
                                }
                            }
                        }
                        catch (Exception ex)
                        {
                            if (DirectoryScanExceptionEvent != null)
                                DirectoryScanExceptionEvent(ex);
                        }
                    }

                }

                private void InitHashtable()
                {

                    CheckSessionCurrentDirectory();

                    if (m_subDirectories != null && m_files != null)
                    {
                        return;
                    }

                    if (m_subDirectories == null)
                    {
                        if (m_caseInsensitive)
                        {
                            m_subDirectories = new Dictionary<string, Directory>(StringComparer.CurrentCultureIgnoreCase); // New Hashtable(CaseInsensitiveComparer.Default)
                        }
                        else
                        {
                            m_subDirectories = new Dictionary<string, Directory>(StringComparer.CurrentCulture); // New Hashtable
                        }
                    }

                    if (m_files == null)
                    {
                        if (m_caseInsensitive)
                        {
                            m_files = new Dictionary<string, System.IO.File>(StringComparer.CurrentCultureIgnoreCase); // New Hashtable(CaseInsensitiveHashCodeProvider.Default, CaseInsensitiveComparer.Default)
                        }
                        else
                        {
                            m_files = new Dictionary<string, System.IO.File>(StringComparer.CurrentCulture); // New Hashtable
                        }
                    }

                    LoadDirectoryItems();

                }

                private bool ParseListLine(string line, ItemInfo info)
                {

                    Match m = MatchingListLine(line, ref info.TimeStamp.Style);

                    if (m == null)
                    {
                        return false;
                    }

                    info.Name = m.Groups["name"].Value;
                    info.FullPath = m_fullPath + info.Name;

                    string dir = m.Groups["dir"].Value;

                    if (dir.Length > 0 && dir != "-")
                    {
                        info.IsDirectory = true;
                        info.FullPath += "/";
                    }
                    else
                    {
                        info.Size = long.Parse(m.Groups["size"].Value);
                    }

                    info.Permission = m.Groups["permission"].Value;
                    info.TimeStamp.RawValue = m.Groups["timestamp"].Value;

                    return true;

                }

                private Match MatchingListLine(string line, ref TimeStampParser.RawDataStyle tsStyle)
                {

                    Match m = m_UnixListLineStyle1.Match(line);

                    if (m.Success)
                    {
                        tsStyle = TimeStampParser.RawDataStyle.UnixDate;
                        return m;
                    }

                    m = m_UnixListLineStyle3.Match(line);

                    if (m.Success)
                    {
                        tsStyle = TimeStampParser.RawDataStyle.UnixDateTime;
                        return m;
                    }

                    m = m_UnixListLineStyle2.Match(line);

                    if (m.Success)
                    {
                        tsStyle = TimeStampParser.RawDataStyle.UnixDate;
                        return m;
                    }

                    m = m_DosListLineStyle1.Match(line);

                    if (m.Success)
                    {
                        tsStyle = TimeStampParser.RawDataStyle.DosDateTime;
                        return m;
                    }

                    m = m_DosListLineStyle2.Match(line);

                    if (m.Success)
                    {
                        tsStyle = TimeStampParser.RawDataStyle.UnixDateTime;
                        return m;
                    }

                    tsStyle = TimeStampParser.RawDataStyle.Undetermined;
                    return null;

                }

                public int CompareTo(object obj)
                {

                    // Directories are sorted by name
                    if (obj is IFile)
                    {
                        return Strings.StrComp(m_name, ((IFile)obj).Name, (m_caseInsensitive ? CompareMethod.Text : CompareMethod.Binary));
                    }
                    else
                    {
                        throw (new ArgumentException("Directory can only be compared to other Directories or Files"));
                    }

                }

            }

        }
    }
}

File: Source/OldCode/Archive/2004 TVA Code Library/Source/Ftp/Directory.vb
' James Ritchie Carroll - 2003
Option Explicit On 

Imports System.IO
Imports System.Text
Imports System.Net.Sockets
Imports System.Text.RegularExpressions

Namespace Ftp

    Public Interface IFile

        ReadOnly Property Parent() As Directory
        ReadOnly Property Name() As String
        ReadOnly Property FullPath() As String
        ReadOnly Property IsFile() As Boolean
        ReadOnly Property IsDirectory() As Boolean
        Property Size() As Long
        Property Permission() As String
        Property TimeStamp() As DateTime

    End Interface

    Friend Class TimeStampParser

        Public Enum RawDataStyle
            UnixDate
            UnixDateTime
            DosDateTime
            Undetermined
        End Enum

        Public RawValue As String
        Public Style As RawDataStyle

        Public Sub New()

            Style = RawDataStyle.Undetermined

        End Sub

        Public Sub New(ByVal RawValue As String, ByVal Style As RawDataStyle)

            Me.RawValue = RawValue
            Me.Style = Style

        End Sub

        Public ReadOnly Property Value() As DateTime
            Get
                If Len(RawValue) > 0 Then
                    Try
                        Select Case Style
                            Case RawDataStyle.UnixDate
                                Return CDate(RawValue)
                            Case RawDataStyle.UnixDateTime
                                Dim sa As String() = RawValue.Split(" "c)
                                Return CDate(sa(0) & " " & sa(1) & " " & Year(Now()) & " " & sa(2))
                            Case RawDataStyle.DosDateTime
                                Return CDate(RawValue)
                            Case Else
                                Return CDate(RawValue)
                        End Select
                    Catch
                        Return DateTime.MinValue
                    End Try
                Else
                    Return DateTime.MinValue
                End If
            End Get
        End Property

    End Class

    Friend Class ItemInfo

        Public Name As String
        Public FullPath As String
        Public Permission As String
        Public IsDirectory As Boolean
        Public Size As Long
        Public TimeStamp As New TimeStampParser()

    End Class

    Public Class Directory

        Implements IFile
        Implements IComparable

        Private m_session As SessionConnected
        Private m_parent As Directory
        Private m_name As String
        Private m_fullPath As String
        Private m_subDirectories As Hashtable
        Private m_files As Hashtable
        Private m_caseInsensitive As Boolean
        Private m_size As Long
        Private m_permission As String
        Private m_timestamp As DateTime

        Private Shared m_UnixListLineStyle1 As New Regex("(?<dir>[\-d])(?<permission>([\-r][\-w][\-xs]){3})\s+\d+\s+\w+\s+\w+\s+(?<size>\d+)\s+(?<timestamp>\w+\s+\d+\s+\d{4})\s+(?<name>.+)")
        Private Shared m_UnixListLineStyle2 As New Regex("(?<dir>[\-d])(?<permission>([\-r][\-w][\-xs]){3})\s+\d+\s+\d+\s+(?<size>\d+)\s+(?<timestamp>\w+\s+\d+\s+\d{4})\s+(?<name>.+)")
        Private Shared m_UnixListLineStyle3 As New Regex("(?<dir>[\-d])(?<permission>([\-r][\-w][\-xs]){3})\s+\d+\s+\d+\s+(?<size>\d+)\s+(?<timestamp>\w+\s+\d+\s+\d+:\d+)\s+(?<name>.+)")
        Private Shared m_DosListLineStyle1 As New Regex("(?<timestamp>\d{2}\-\d{2}\-\d{2}\s+\d{2}:\d{2}[Aa|Pp][mM])\s+(?<dir>\<\w+\>){0,1}(?<size>\d+){0,1}\s+(?<name>.+)")                     ' IIS FTP Service
        Private Shared m_DosListLineStyle2 As New Regex("(?<dir>[\-d])(?<permission>([\-r][\-w][\-xs]){3})\s+\d+\s+\w+\s+\w+\s+(?<size>\d+)\s+(?<timestamp>\w+\s+\d+\s+\d+:\d+)\s+(?<name>.+)") ' IIS FTP Service in Unix Mode

        Public Event DirectoryListLineScan(ByVal Line As String)

        Friend Sub New(ByVal s As SessionConnected, ByVal CaseInsensitive As Boolean, ByVal fullPath As String)

            m_session = s
            m_parent = Nothing
            m_caseInsensitive = CaseInsensitive

            If Right(fullPath, 1) = "/" Then fullPath = Left(fullPath, Len(fullPath) - 1)

            If Len(fullPath) = 0 Then
                m_name = m_fullPath = "/"
            Else
                Dim directories As String() = fullPath.Split("/"c)
                m_name = directories(directories.Length - 1)
                m_fullPath = fullPath & "/"
            End If

        End Sub

        Friend Sub New(ByVal s As SessionConnected, ByVal parent As Directory, ByVal CaseInsensitive As Boolean, ByVal info As ItemInfo)

            m_session = s
            m_parent = parent
            m_caseInsensitive = CaseInsensitive

            If Len(info.Name) > 0 Then
                m_name = info.Name
                If parent Is Nothing Then
                    m_fullPath = m_name & "/"
                Else
                    m_fullPath = parent.FullPath & m_name & "/"
                End If
            Else
                m_name = m_fullPath = "/"
            End If

            m_size = info.Size
            m_permission = info.Permission
            m_timestamp = info.TimeStamp.Value

        End Sub

        Public Property CaseInsensitive() As Boolean
            Get
                Return m_caseInsensitive
            End Get
            Set(ByVal Value As Boolean)
                m_caseInsensitive = Value
                Refresh()
            End Set
        End Property

        Public ReadOnly Property Name() As String Implements IFile.Name
            Get
                Return m_name
            End Get
        End Property

        Public ReadOnly Property FullPath() As String Implements IFile.FullPath
            Get
                Return m_fullPath
            End Get
        End Property

        Public ReadOnly Property IsFile() As Boolean Implements IFile.IsFile
            Get
                Return False
            End Get
        End Property

        Public ReadOnly Property IsDirectory() As Boolean Implements IFile.IsDirectory
            Get
                Return True
            End Get
        End Property

        Public Property Size() As Long Implements IFile.Size
            Get
                Return m_size
            End Get
            Set(ByVal Value As Long)
                m_size = Value
            End Set
        End Property

        Public Property Permission() As String Implements IFile.Permission
            Get
                Return m_permission
            End Get
            Set(ByVal Value As String)
                m_permission = Value
            End Set
        End Property

        Public Property TimeStamp() As DateTime Implements IFile.TimeStamp
            Get
                Return m_timestamp
            End Get
            Set(ByVal Value As DateTime)
                m_timestamp = Value
            End Set
        End Property

        Public ReadOnly Property Parent() As Directory Implements IFile.Parent
            Get
                If StrComp(m_fullPath, m_session.RootDirectory.m_fullPath, IIf(m_caseInsensitive, CompareMethod.Text, CompareMethod.Binary)) = 0 Then
                    Return Nothing
                End If

                ' If we don't have a reference to parent directory, we try to derive it...
                If m_parent Is Nothing Then
                    CheckSessionCurrentDirectory()

                    Dim parentPath As New StringBuilder
                    Dim fullPath = m_session.ControlChannel.PWD()

                    If Right(fullPath, 1) <> "/" Then fullPath &= "/"

                    Dim paths As String() = fullPath.Split("/")
                    Dim i As Integer

                    For i = 0 To paths.Length - 2
                        If Len(paths(i)) = 0 Then
                            parentPath.Append("/")
                        Else
                            parentPath.Append(paths(i))
                            parentPath.Append("/")
                        End If
                    Next i

                    Dim parentDir As New Directory(m_session, m_caseInsensitive, parentPath.ToString())

                    If StrComp(parentDir.m_fullPath, m_session.RootDirectory.m_fullPath, IIf(m_caseInsensitive, CompareMethod.Text, CompareMethod.Binary)) = 0 Then
                        m_parent = m_session.RootDirectory
                    Else
                        m_parent = parentDir
                    End If
                End If

                Return m_parent
            End Get
        End Property

        Public ReadOnly Property SubDirectories() As ICollection
            Get
                InitHashtable()
                Return m_subDirectories.Values
            End Get
        End Property

        Public ReadOnly Property Files() As ICollection
            Get
                InitHashtable()
                Return m_files.Values
            End Get
        End Property

        Public Function FindFile(ByVal fileName As String) As File

            InitHashtable()

            Return CType(m_files(fileName), File)

        End Function

        Public Overloads Function FindSubDirectory(ByVal dirName As String) As Directory

            InitHashtable()

            Return CType(m_subDirectories(dirName), Directory)

        End Function

        Public Overloads Sub PutFile(ByVal localFile As String)

            PutFile(localFile, Nothing)

        End Sub

        Public Overloads Sub PutFile(ByVal localFile As String, ByVal remoteFile As String)

            CheckSessionCurrentDirectory()

            Dim fi As New FileInfo(localFile)

            If remoteFile Is Nothing Then
                remoteFile = fi.Name
            End If

            Dim transfer As New FileTransferer(Me, localFile, remoteFile, fi.Length, TransferDirection.Upload)

            transfer.StartTransfer()

        End Sub

        Public Overloads Sub GetFile(ByVal remoteFile As String)

            GetFile(remoteFile, remoteFile)

        End Sub

        Public Overloads Sub GetFile(ByVal localFile As String, ByVal remoteFile As String)

            InitHashtable()

            Dim file As File = CType(m_files(remoteFile), File)

            If file Is Nothing Then
                Throw New FileNotFoundException(remoteFile)
            End If

            Dim transfer As New FileTransferer(Me, localFile, remoteFile, file.Size, TransferDirection.Download)

            transfer.StartTransfer()

        End Sub

        Public Overloads Sub BeginPutFile(ByVal localFile As String)

            BeginPutFile(localFile, Nothing)

        End Sub

        Public Overloads Sub BeginPutFile(ByVal localFile As String, ByVal remoteFile As String)

            CheckSessionCurrentDirectory()

            Dim fi As New FileInfo(localFile)

            If remoteFile Is Nothing Then
                remoteFile = fi.Name
            End If

            Dim transfer As New FileTransferer(Me, localFile, remoteFile, fi.Length, TransferDirection.Upload)

            transfer.StartAsyncTransfer()

        End Sub

        Public Overloads Sub BeginGetFile(ByVal remoteFile As String)

            BeginGetFile(remoteFile, remoteFile)

        End Sub

        Public Overloads Sub BeginGetFile(ByVal localFile As String, ByVal remoteFile As String)

            InitHashtable()

            Dim file As File = CType(m_files(remoteFile), File)

            If file Is Nothing Then
                Throw New FileNotFoundException(remoteFile)
            End If

            Dim transfer As New FileTransferer(Me, localFile, remoteFile, file.Size, TransferDirection.Download)

            transfer.StartAsyncTransfer()

        End Sub

        Public Sub RemoveFile(ByVal fileName As String)

            CheckSessionCurrentDirectory()

            m_session.ControlChannel.DELE(fileName)

            m_files.Remove(fileName)

        End Sub

        Public Sub RemoveSubDir(ByVal dirName As String)

            CheckSessionCurrentDirectory()

            m_session.ControlChannel.RMD(dirName)

            m_subDirectories.Remove(dirName)

        End Sub

        Public Function CreateFile(ByVal newFileName As String) As File

            Dim stream As DataStream = CreateFileStream(newFileName)

            stream.Close()

            Return CType(m_files(newFileName), File)

        End Function

        Public Function CreateFileStream(ByVal newFileName As String) As OutputDataStream

            InitHashtable()

            Dim stream As DataStream = m_session.ControlChannel.GetPassiveDataStream(TransferDirection.Upload)

            Try
                m_session.ControlChannel.STOR(newFileName)

                Dim newFile As New File(Me, newFileName)

                m_files(newFileName) = newFile

                Return CType(stream, OutputDataStream)
            Catch
                stream.Close()
                Throw
            End Try

        End Function

        Public Sub Refresh()

            ClearItems()
            InitHashtable()

        End Sub

        Friend Sub ClearItems()

            m_subDirectories = Nothing
            m_files = Nothing

        End Sub

        Friend ReadOnly Property Session() As SessionConnected
            Get
                Return m_session
            End Get
        End Property

        Friend Sub CheckSessionCurrentDirectory()

            If m_session.CurrentDirectory.m_fullPath <> m_fullPath Then
                Throw New InvalidOperationException(m_fullPath & " is not current directory.")
            End If

        End Sub

        Private Sub LoadDirectoryItems()

            If Not m_session.CurrentDirectory Is Me Then
                Throw New InvalidOperationException(m_name & " is not current active directory")
            End If

            Dim lineQueue As Queue = m_session.ControlChannel.List(False)
            Dim info As ItemInfo
            Dim line As String

            For Each line In lineQueue
                ' We allow users to inspect FTP lineQueue if desired...
                RaiseEvent DirectoryListLineScan(line)

                info = New ItemInfo
                If ParseListLine(line, info) Then
                    If info.IsDirectory Then
                        m_subDirectories.Add(info.Name, New Directory(m_session, Me, m_caseInsensitive, info))
                    Else
                        m_files.Add(info.Name, New File(Me, info))
                    End If
                End If
            Next

        End Sub

        Private Sub InitHashtable()

            CheckSessionCurrentDirectory()

            If Not m_subDirectories Is Nothing AndAlso Not m_files Is Nothing Then
                Return
            End If

            If m_subDirectories Is Nothing Then
                If m_caseInsensitive Then
                    m_subDirectories = New Hashtable(CaseInsensitiveHashCodeProvider.Default, CaseInsensitiveComparer.Default)
                Else
                    m_subDirectories = New Hashtable
                End If
            End If

            If m_files Is Nothing Then
                If m_caseInsensitive Then
                    m_files = New Hashtable(CaseInsensitiveHashCodeProvider.Default, CaseInsensitiveComparer.Default)
                Else
                    m_files = New Hashtable
                End If
            End If

            LoadDirectoryItems()

        End Sub

        Private Function ParseListLine(ByVal line As String, ByVal info As ItemInfo) As Boolean

            Dim m As Match = MatchingListLine(line, info.TimeStamp.Style)

            If m Is Nothing Then
                Return False
            End If

            info.Name = m.Groups("name").Value
            info.FullPath = m_fullPath & info.Name

            Dim dir As String = m.Groups("dir").Value

            If Len(dir) > 0 AndAlso dir <> "-" Then
                info.IsDirectory = True
                info.FullPath &= "/"
            Else
                info.Size = Long.Parse(m.Groups("size").Value)
            End If

            info.Permission = m.Groups("permission").Value
            info.TimeStamp.RawValue = m.Groups("timestamp").Value

            Return True

        End Function

        Private Function MatchingListLine(ByVal line As String, ByRef tsStyle As TimeStampParser.RawDataStyle) As Match

            Dim m As Match = m_UnixListLineStyle1.Match(line)

            If m.Success Then
                tsStyle = TimeStampParser.RawDataStyle.UnixDate
                Return m
            End If

            m = m_UnixListLineStyle3.Match(line)

            If m.Success Then
                tsStyle = TimeStampParser.RawDataStyle.UnixDateTime
                Return m
            End If

            m = m_UnixListLineStyle2.Match(line)

            If m.Success Then
                tsStyle = TimeStampParser.RawDataStyle.UnixDate
                Return m
            End If

            m = m_DosListLineStyle1.Match(line)

            If m.Success Then
                tsStyle = TimeStampParser.RawDataStyle.DosDateTime
                Return m
            End If

            m = m_DosListLineStyle2.Match(line)

            If m.Success Then
                tsStyle = TimeStampParser.RawDataStyle.UnixDateTime
                Return m
            End If

            tsStyle = TimeStampParser.RawDataStyle.Undetermined
            Return Nothing

        End Function

        Public Function CompareTo(ByVal obj As Object) As Integer Implements System.IComparable.CompareTo

            ' Directories are sorted by name
            If TypeOf obj Is IFile Then
                Return StrComp(m_name, DirectCast(obj, IFile).Name, IIf(m_caseInsensitive, CompareMethod.Text, CompareMethod.Binary))
            Else
                Throw New ArgumentException("Directory can only be compared to other Directories or Files")
            End If

        End Function

    End Class

End Namespace